include/RED4ext/Containers/DynArray.hpp

```diff
@@ -0,0 +1,550 @@
+    }
+
+    DynArray(uint32_t aSize, Memory::IAllocator* aAllocator = nullptr)
+        : DynArray(aAllocator)
+    {
+        Resize(aSize);
+    }
+
+    DynArray(const DynArray& aOther)
+        : DynArray(aOther.GetAllocator())
+    {
+        Assign(aOther.begin(), aOther.end());
+    }
+
+    DynArray(DynArray&& aOther) noexcept
+        : entries(aOther.Data())
+        , size(aOther.Size())
+        , capacity(aOther.Capacity())
+    {
+        aOther.entries = *reinterpret_cast<Pointer*>(aOther.GetAllocator());
+        aOther.capacity = 0;
+        aOther.size = 0;
+    }
+
+    template<ContainerType TContainer>
+    DynArray(const TContainer& aContainer, Memory::IAllocator* aAllocator = nullptr)
+        : DynArray(aAllocator)
+    {
+        Assign(aContainer.begin(), aContainer.end());
+    }
+
+    template<ContainerType TContainer>
+    DynArray(TContainer&& aContainer, Memory::IAllocator* aAllocator = nullptr)
+        : DynArray(aAllocator)
+    {
+        Assign(std::make_move_iterator(aContainer.begin()), std::make_move_iterator(aContainer.end()));
+    }
+
+    ~DynArray()
+    {
+        Clear();
+        auto allocator = *reinterpret_cast<Pointer*>(GetAllocator());
+        reinterpret_cast<Memory::IAllocator*>(&allocator)->Free(entries);
+        entries = allocator;
+        capacity = 0;
+    }
+
+    DynArray& operator=(const DynArray& aOther)
+    {
+        if (this != std::addressof(aOther))
+        {
+            Assign(aOther.begin(), aOther.end());
+        }
+        return *this;
+    }
+
+    DynArray& operator=(DynArray&& aOther)
+    {
+        if (this != std::addressof(aOther))
+        {
+            Assign(std::make_move_iterator(aOther.begin()), std::make_move_iterator(aOther.end()));
+        }
+        return *this;
+    }
+
+    template<ContainerType TContainer>
+    DynArray& operator=(const TContainer& aOther)
+    {
+        Assign(aOther.begin(), aOther.end());
+        return *this;
+    }
+
+    template<ContainerType TContainer>
+    DynArray& operator=(TContainer&& aOther)
+    {
+        if (this != std::addressof(aOther))
+        {
+            Assign(std::make_move_iterator(aOther.begin()), std::make_move_iterator(aOther.end()));
+        }
+
+        return *this;
+    }
+
+    constexpr Reference operator[](SizeType aPos)
+    {
+        assert(aPos < Size());
+        return Data()[aPos];
+    }
+
+    constexpr ConstReference operator[](SizeType aPos) const
+    {
+        assert(aPos < Size());
+        return Data()[aPos];
+    }
+
+    template<typename TIterator>
+    constexpr void Assign(TIterator aBegin, TIterator aEnd)
+    {
+        SizeType newSize = static_cast<SizeType>(std::distance(aBegin, aEnd));
+        assert(newSize >= 0);
+
+        Clear();
+        Resize(newSize);
+
+        if constexpr (std::is_trivially_copyable_v<ValueType>)
+        {
+            auto begin = *aBegin;
+            std::memcpy(Data(), &begin, newSize * sizeof(ValueType));
+        }
```

Comment:
Never do a `memcpy` or `memmove`. You might have pointers in the object that can point to itself, or to another variable in the object. Doing a `copy` you will have undefined behavior.